{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "import pickle\n",
    "\n",
    "SEED = 42\n",
    "\n",
    "tissue_param = {\"lung\": {\n",
    "                \"name\": \"Lung\", \n",
    "                \"age_data\": r\"../../metadata/lung_annotation_meth.csv\",\n",
    "                \"test_data\": \"../../metadata/lung_test_metadata.csv\"\n",
    "            }, \n",
    "            \"ovary\": {\n",
    "                \"name\": \"Ovary\", \n",
    "                \"age_data\": r\"../../metadata/ovary_annotation_meth.csv\",\n",
    "                \"test_data\": \"../../metadata/ovary_test_metadata.csv\"\n",
    "            },\n",
    "        }\n",
    "\n",
    "\n",
    "\n",
    "def load_lung_data() -> pd.DataFrame:\n",
    "    return pd.read_csv(\"../../data/methylation_lung.csv\")\n",
    "\n",
    "def load_ovary_data() -> pd.DataFrame:\n",
    "    return pd.read_csv(\"../../data/methylation_ovary.csv\")\n",
    "\n",
    "\n",
    "tissues_to_run = [\"lung\", \"ovary\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def split_in_train_test(meth, age_data, test_set):\n",
    "    \"\"\"Split the dataset into train and test\"\"\"\n",
    "    #Metadata\n",
    "    metadata_test = age_data.loc[age_data[\"tissue_sample_id\"].isin(test_set[\"sample_id\"])]\n",
    "    metadata_train = age_data.loc[~age_data[\"tissue_sample_id\"].isin(test_set[\"sample_id\"])]\n",
    "    #Age data\n",
    "    y_test = list(metadata_test[\"AGE\"])\n",
    "    y_train = list(metadata_train[\"AGE\"])\n",
    "    #Methylation data\n",
    "    meth_t = meth.transpose()\n",
    "    X_train = meth_t.loc[metadata_train.Sample_ID]\n",
    "    X_test = meth_t.loc[metadata_test.Sample_ID]\n",
    "\n",
    "    return(X_train, X_test, y_train, y_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Acer\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\sklearn\\model_selection\\_split.py:725: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.\n",
      "  warnings.warn(\n",
      "c:\\Users\\Acer\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\sklearn\\model_selection\\_split.py:725: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "for tissue in tissues_to_run:\n",
    "    res_dir = f\"../../results/3.feature_selection_multimodal/{tissue}\"\n",
    "    if not os.path.exists(res_dir): \n",
    "        os.mkdir(res_dir)\n",
    "\n",
    "    if tissue == \"lung\":\n",
    "        meth_data = load_lung_data()\n",
    "        tissue_name = \"Lung\"\n",
    "    elif tissue == \"ovary\": \n",
    "        meth_data = load_ovary_data()\n",
    "        tissue_name = \"Ovary\"\n",
    "    else: \n",
    "        print(\"Error.. No tissue with name, \", tissue)\n",
    "        next\n",
    "\n",
    "    # Load age data\n",
    "    age_data = tissue_param[tissue][\"age_data\"]\n",
    "    age_data = pd.read_csv(age_data)\n",
    "\n",
    "    # Load test set\n",
    "    test_data = tissue_param[tissue][\"test_data\"]\n",
    "    test_set = pd.read_csv(test_data)\n",
    "    \n",
    "    #Filter for complete samples (with all data types)\n",
    "    multi_modal_table = \"../../metadata/sample_ids_multiomics_updated_tl_data.csv\"\n",
    "    multi_modal_table = pd.read_csv(multi_modal_table)\n",
    "\n",
    "    filtered_multi_modal_table = multi_modal_table[multi_modal_table['tissue'] == tissue_name]\n",
    "    complete_samples = filtered_multi_modal_table[(filtered_multi_modal_table[\"metadata\"]) == 1 & (filtered_multi_modal_table[\"gene_expression\"]) & (filtered_multi_modal_table[\"metilation\"]) &  (filtered_multi_modal_table[\"telemore\"])]\n",
    "\n",
    "\n",
    "    #Load Athos samples (the ones he was able to process)\n",
    "    if tissue == \"lung\": \n",
    "        hist_data = pickle.load(open(\"../../data/features_histology/lung/lung_features_mean256_cls4k.pkl\", \"rb\"))\n",
    "    if tissue == \"ovary\": \n",
    "        hist_data = pickle.load(open(\"../../data/features_histology/ovary/ovary_features_mean256_cls4k.pkl\", \"rb\"))\n",
    "    # Remove samples with histology that were not processed\n",
    "    complete_samples = complete_samples[complete_samples.sample_id.isin(hist_data.index)]\n",
    "\n",
    "    #Divide in train and test set\n",
    "    age_data_multi_modal = age_data.loc[age_data[\"tissue_sample_id\"].isin(complete_samples[\"sample_id\"])]\n",
    "    X_train, X_test, y_train, y_test = split_in_train_test(meth_data, age_data_multi_modal, test_set)\n",
    "    \n",
    "    #Define folds for CV\n",
    "    n_folds = 5\n",
    "\n",
    "    age_bins = np.arange(18, 76, 5)\n",
    "    y_bins = pd.cut(y_train, age_bins, labels=False)\n",
    "    skf_gen = StratifiedKFold(n_folds, shuffle = True, random_state = SEED).split(X_train, y_bins)\n",
    "\n",
    "    folds = [[t[0], t[1]] for t in skf_gen]\n",
    "\n",
    "    # Save folds to a text file\n",
    "    for i in range(len(folds)): \n",
    "        fold = folds[i]\n",
    "        train_data = fold[0]\n",
    "        test_data = fold[1] \n",
    "\n",
    "        train_samples = X_train.index[train_data]\n",
    "        test_samples = X_train.index[test_data]\n",
    "        \n",
    "        train_data = pd.DataFrame({\"index\": train_data, \"sample\": train_samples})\n",
    "        test_data = pd.DataFrame({\"index\": test_data, \"sample\": test_samples})\n",
    "        \n",
    "        # Save train and test data to separate CSV files\n",
    "        train_data.to_csv(f'../../results/3.feature_selection_multimodal/{tissue}/fold_{i}_train.csv', index=False)\n",
    "        test_data.to_csv(f'../../results/3.feature_selection_multimodal/{tissue}/fold_{i}_test.csv', index=False)\n",
    "        \n",
    "        # Save exclusive DNA methyl data (used during feature selection)\n",
    "        exclusive_methylation_samples = age_data[~age_data[\"tissue_sample_id\"].isin(complete_samples.sample_id)]\n",
    "        exclusive_methylation_samples.to_csv(f'../../results/3.feature_selection_multimodal/{tissue}/exclusive_methylation_train.csv', index=False)\n",
    "\n",
    "        # Save exclusive Gene expression data (used during feature selection)\n",
    "        gexp_metadata = pd.read_csv(f\"../../metadata/gene_expression_metadata/metadata_{tissue}.tsv\", sep =\"\\t\")\n",
    "        gexp_metadata_exclusive = gexp_metadata[~gexp_metadata[\"tissue_sample_id\"].isin(complete_samples.sample_id)]\n",
    "        gexp_metadata_exclusive.to_csv(f'../../results/3.feature_selection_multimodal/{tissue}/exclusive_gene_expression_train.csv', index=False)\n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
